### PR TITLE
Add a check for stabby lambda parameters parentheses.

### PR DESCRIPTION
**Pull request by hmadison**
Adds in a cop to catch cases like `->x,y,z { z + x + y }` and correct them to `->(x,y,z) { z + x + y }`. It can also be configured to perform the reverse, catch `->(x,y,z) { z + x + y }` and correct to  `->x,y,z { z + x + y }`.  Resolves issue [#2036](https://github.com/bbatsov/rubocop/issues/2036).

### CODE REVIEW

File: lib/rubocop/cop/style/stabby_lambda_parentheses.rb [hunk 1]
```diff
@@ -0,0 +1,92 @@
+# encoding: utf-8
+module RuboCop
+  module Cop
+    module Style
+      # Check for parentheses around stabby lambda arguments
```
Comment (bbatsov) [inline lib/rubocop/cop/style/stabby_lambda_parentheses.rb]: This should be extended a bit to feature some code examples. See other cops for details.

File: lib/rubocop/cop/style/stabby_lambda_parentheses.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+        def autocorrect(node)
+          if style == :require_parentheses
+            missing_parentheses_corrector(node)
+          elsif style == :require_no_parentheses
+            unwanted_parentheses_corrector(node)
+          end
+        end
+
+        def missing_parentheses(node)
```
Comment (bbatsov) [inline lib/rubocop/cop/style/stabby_lambda_parentheses.rb]: This method and the one after it should be private as well.

File: lib/rubocop/cop/style/stabby_lambda_parentheses.rb [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+          add_offense(node_args(node), :expression,
+                      'Use with parentheses when there are parameters.') do
```
Comment (bbatsov) [inline lib/rubocop/cop/style/stabby_lambda_parentheses.rb]: Extract this message to a constant. I'd also rephrase it to something like `Wrap stabby lambda arguments with parentheses.`.

File: lib/rubocop/cop/style/stabby_lambda_parentheses.rb [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            opposite_style_detected
+          end
+        end
+
+        def unwanted_parentheses(node)
+          add_offense(node_args(node), :expression,
+                      'Use without parentheses when there are parameters.') do
```
Comment (bbatsov) [inline lib/rubocop/cop/style/stabby_lambda_parentheses.rb]: Same remarks here.

File: lib/rubocop/cop/style/stabby_lambda_parentheses.rb [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        def unwanted_parentheses_corrector(node)
+          lambda do |corrector|
+            args_loc = node_args(node).loc
+
+            corrector.replace(args_loc.begin, '')
+            corrector.remove(args_loc.end)
+          end
+        end
+
+        def valid_node?(node)
```
Comment (bbatsov) [inline lib/rubocop/cop/style/stabby_lambda_parentheses.rb]: This name is a bit unclear. Perhaps `lambda_with_args?`?

File: lib/rubocop/cop/style/stabby_lambda_parentheses.rb [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+          lambda_node?(node) && args?(node)
+        end
+
+        def lambda_node?(node)
+          _, call = node.to_a
+          call == :lambda
+        end
+
+        def node_args(node)
+          node.parent.children.detect(&:args_type?)
```
Comment (bbatsov) [inline lib/rubocop/cop/style/stabby_lambda_parentheses.rb]: I'd rather use node destructuring with * to get the args.

File: lib/rubocop/cop/style/stabby_lambda_parentheses.rb [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+            corrector.remove(args_loc.end)
+          end
+        end
+
+        def valid_node?(node)
+          lambda_node?(node) && args?(node)
+        end
+
+        def lambda_node?(node)
+          _, call = node.to_a
```
Comment (bbatsov) [inline lib/rubocop/cop/style/stabby_lambda_parentheses.rb]: Just use splatting here. We also generally don't use `_`, but `_something` for greater clarity.

File: lib/rubocop/cop/style/stabby_lambda_parentheses.rb [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+          call == :lambda
+        end
+
+        def node_args(node)
+          node.parent.children.detect(&:args_type?)
+        end
+
+        def args?(node)
+          args = node_args(node)
+          args.children.count > 0
```
Comment (bbatsov) [inline lib/rubocop/cop/style/stabby_lambda_parentheses.rb]: Guess you're not receiving the right node here if you need its children.

File: spec/rubocop/cop/style/stabby_lambda_parentheses_spec.rb [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    it 'registers an offense for a stabby lambda without parentheses' do
+      inspect_source(cop, '->a,b,c { a + b + c }')
+      expect(cop.offenses.size).to eq(1)
+      expect(cop.messages)
+        .to eq(['Use with parentheses when there are parameters.'])
+    end
+
+    it 'does not register an offense for a stabby lambda with parentheses' do
+      inspect_source(cop, '->(a,b,c) { a + b + c }')
+      expect(cop.offenses.size).to eq(0)
```
Comment (bbatsov) [inline spec/rubocop/cop/style/stabby_lambda_parentheses_spec.rb]: We generally favour the use of `to be_empty` in such scenarios.

File: lib/rubocop/cop/style/stabby_lambda_parentheses.rb (hunk 8 above)
Comment (hmadison) [inline lib/rubocop/cop/style/stabby_lambda_parentheses.rb]: I'm doing a similar check to [`MethodDefParentheses#arguments?`](https://github.com/bbatsov/rubocop/blob/master/lib/rubocop/cop/style/method_def_parentheses.rb#L68-L70).  Is there a better way to test the args node for the presence of arguments?

**Comment by hmadison** [comment]
Rebased and updated.

File: CHANGELOG.md [hunk 10]
```diff
@@ -1608,4 +1612,8 @@
 [@imtayadeway]: https://github.com/imtayadeway
 [@mrfoto]: https://github.com/mrfoto
 [@karreiro]: https://github.com/karreiro
+<<<<<<< HEAD
```
Comment (bbatsov) [inline CHANGELOG.md]: Merge conflict :-)

File: lib/rubocop/cop/style/stabby_lambda_parentheses.rb (hunk 8 above)
Comment (bbatsov) [inline lib/rubocop/cop/style/stabby_lambda_parentheses.rb]: I was mostly referring to the fact that the `children` are the actual arguments. I can live with the current version of the code.

**Comment by bbatsov** [comment]
You should also squash the two commits together, drop the `.` from the commit message's title and prefix the message with `[Fix #2036]` (as noted in CONTRIBUTING.md).

**Comment by hmadison** [comment]
Squashed and updated.

File: lib/rubocop/cop/style/stabby_lambda_parentheses.rb [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+        MSG_REQUIRE = 'Wrap stabby lambda arguments with parentheses.'
+        MSG_NO_REQUIRE = 'Do not wrap stabby lambda arguments with parentheses.'
+
+        include ConfigurableEnforcedStyle
+
+        def on_send(node)
+          return unless lambda_with_args?(node)
+
+          if style == :require_parentheses
+            if !parentheses?(node)
```
Comment (rrosenblum) [inline lib/rubocop/cop/style/stabby_lambda_parentheses.rb]: Should this if statement be inverted to use the positive form?

```ruby
if parentheses?(node)
  correct_style_detected
else
  missing_parentheses(node)
end
```

File: lib/rubocop/cop/style/stabby_lambda_parentheses.rb [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+            unwanted_parentheses(node)
+          else
+            correct_style_detected
+          end
+        end
+
+        def autocorrect(node)
+          if style == :require_parentheses
+            missing_parentheses_corrector(node)
+          elsif style == :require_no_parentheses
```
Comment (rrosenblum) [inline lib/rubocop/cop/style/stabby_lambda_parentheses.rb]: Should this use `else` instead of `elsif`? There are only 2 configurable styles.

Comment (hmadison) [inline lib/rubocop/cop/style/stabby_lambda_parentheses.rb]: It could be. I feel though its important to keep some indication between the style and the corrector. Would a case statement be better?

File: lib/rubocop/cop/style/stabby_lambda_parentheses.rb (hunk 11 above)
Comment (hmadison) [inline lib/rubocop/cop/style/stabby_lambda_parentheses.rb]: Yes, ill update this.

File: CHANGELOG.md [hunk 13]
```diff
@@ -11,10 +11,15 @@
 * [#2121](https://github.com/bbatsov/rubocop/issues/2121): Allow space before values in hash literals in `Style/ExtraSpacing` to avoid correction conflict. ([@jonas054][])
 * [#2241](https://github.com/bbatsov/rubocop/issues/2241): Read cache in binary format. ([@jonas054][])
 
+
 ### Changes
 
 * [#2248](https://github.com/bbatsov/rubocop/issues/2248): Allow block-pass in `Style/AutoResourceCleanup`. ([@lumeet][])
 
+### New features
```
Comment (bbatsov) [inline CHANGELOG.md]: I just released 0.34.2, so you should rebase on top of the current `master` and move this section to its proper location. Once this is done, we can finally merge the PR.

Comment (hmadison) [inline CHANGELOG.md]: Rebased and updated.

File: CHANGELOG.md [hunk 14]
```diff
@@ -18,6 +22,7 @@
 * [#2255](https://github.com/bbatsov/rubocop/issues/2255): Refine the offense highlighting for `Style/SymbolProc`. ([@bbatsov][])
 * [#2260](https://github.com/bbatsov/rubocop/pull/2260): Make `Exclude` in `.rubocop_todo.yml` work when running from a subdirectory. ([@jonas054][])
 
+
```
Comment (bbatsov) [inline CHANGELOG.md]: You should probably remove the extra blank line. Not sure how it got here, but I guess it was by accident.

Comment (hmadison) [inline CHANGELOG.md]: Fixed.

**Comment by bbatsov** [comment]
:+1: Thanks!
